faucet/src/Dscp/Faucet/Web/Logic.hs

```diff
@@ -0,0 +1,67 @@
+-- | Faucet backend logic.
+
+module Dscp.Faucet.Web.Logic
+    ( faucetTransferMoneyTo
+    ) where
+
+import Control.Lens (at, views, (<<.=))
+import Loot.Base.HasLens (HasCtx, lensOf)
+import Serokell.Util (modifyTVarS)
+import UnliftIO.MVar (withMVar)
+
+import Dscp.Core
+import Dscp.Crypto
+import Dscp.Faucet.Launcher.Marker
+import Dscp.Faucet.Launcher.Mode
+import Dscp.Faucet.Launcher.Params
+import Dscp.Faucet.Variables
+import Dscp.Faucet.Web.Error
+import Dscp.Resource.Keys
+import Dscp.Witness.Web.Client
+import Dscp.Witness.Web.Types
+
+-- | Atomcally remember this address as being gifted or throw if it is not
+-- its first time.
+ensureFirstGift
+    :: (MonadIO m, MonadThrow m, HasCtx ctx m '[GiftedAddresses])
+    => Address -> m ()
+ensureFirstGift addr = do
+    giftedAddresses <- view (lensOf @GiftedAddresses)
+    prevVal <- atomically . modifyTVarS giftedAddresses $
+        at addr <<.= Just ()
+    when (isJust prevVal) $
+        throwM AddressAlreadyGifted
```

Comment:
How do these translate to user errors? Are they returned to the frontend?

Reply:
Yes, they are `show`n and passed to json